yo1k/sandbox/exercises/CSC_algorithms/sorting.py

```diff
@@ -0,0 +1,131 @@
+from collections.abc import MutableSequence
+from typing import Any, Optional
+
+
+def swap(array: MutableSequence[Any], frst_idx: int, snd_idx: int) -> MutableSequence[Any]:
+    """Swaps to elements in a list."""
+    array[frst_idx], array[snd_idx] = array[snd_idx], array[frst_idx]
+    return array
+
+
+def selection_sort(array: MutableSequence[Any]) -> MutableSequence[Any]:
+    """Sorting by selection.
+
+    Algorithm in-place, complexity: T(n) = O(n**2).
+    """
+    size: int = len(array)
+    for i in range(size):
+        min_idx: int = -1
+        for j in range(i, size):
+            if min_idx == -1 or array[j] <= array[min_idx]:
+                min_idx = j
+        swap(array, min_idx, i)
+    return array
+
+
+def insertion_sort(array: MutableSequence[Any]) -> MutableSequence[Any]:
+    """Sorting by inserts.
+
+    Algorithm in-place, complexity: T(n) = O(n**2).
+    """
+    for i in range(len(array)):
+        j: int = i
+        while j > 0 and array[j - 1] > array[j]:
+            swap(array, j-1, j)
+            j -= 1
+    return array
+
+
+def merge_sorted_arrays(
+        array_left: MutableSequence[Any],
+        array_right: MutableSequence[Any],
+        merged_array: Optional[MutableSequence[Any]] = None) -> MutableSequence[Any]:
+    """Merges two sorted arrays in ascending order."""
+    merged_array = [] if merged_array is None else merged_array
+    l_idx: int = 0
+    r_idx: int = 0
+    l_end: int = len(array_left)
+    r_end: int = len(array_right)
```

Comment:
These variables represent the lengths of the lists, not end indices. Naming them `*_end` suggests that they contain indices. I propose renaming them to `*_len`.

Reply:
Done.